compiler/rustc_typeck/src/astconv/mod.rs

```diff
@@ -1444,19 +1444,26 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
                 // Verify that `dummy_self` did not leak inside default type parameters.  This
                 // could not be done at path creation, since we need to see through trait aliases.
                 let mut missing_type_params = vec![];
+                let mut references_self = false;
                 let generics = tcx.generics_of(trait_ref.def_id);
                 let substs: Vec<_> = trait_ref
                     .substs
                     .iter()
                     .enumerate()
                     .skip(1) // Remove `Self` for `ExistentialPredicate`.
                     .map(|(index, arg)| {
-                        if let ty::GenericArgKind::Type(ty) = arg.unpack()
-                            && ty == dummy_self
-                        {
-                            let param = &generics.params[index];
-                            missing_type_params.push(param.name);
-                            tcx.ty_error().into()
+                        if let ty::GenericArgKind::Type(ty) = arg.unpack() {
+                            debug!(?ty);
+                            if ty == dummy_self {
+                                let param = &generics.params[index];
+                                missing_type_params.push(param.name);
+                                tcx.ty_error().into()
+                            } else if ty.has_fresh_types() {
```

Comment:
maybe instead use `ty.walk().any(|arg| arg == dummy_self.into())` :thinking: this feels a lot more clear than checking for `HAS_TY_FRESH`